fix(normalization): Light normalize stacktrace before filtering [ISSUE-1574]

iker-barriocanal opened the pull request:
Browser extension filters require the normalization of the last frame of
the stacktrace of the first exception of the event.

Currently, this isn't part of the light normalization, which makes the
filter useless and events originating in browser extensions get to Sentry.

iker-barriocanal committed:
fix(filters): Light normalize stacktrace before filtering

Browser extension filters require the normalization of the last frame of
the stacktrace of the first exception of the event.

Currently, this isn't part of the light normalization, which makes the
filter useless and events coming from browser extensions get to Sentry.

iker-barriocanal committed:
Add integration test

iker-barriocanal committed:
Add changelog entry

iker-barriocanal commented on relay-general/src/store/normalize.rs:
```diff
@@ -555,6 +584,7 @@ pub fn light_normalize_event(
         )?; // Timestamps are core in the metrics extraction
         normalize_event_tags(event)?; // Tags are added to every metric
         normalize_exceptions(event)?; // Browser extension filters look at the stacktrace
+        light_normalize_stacktraces(event)?;
```
`light_normalize_stacktraces` doesn't replace `normalize_exceptions`, as it works with stacktraces. The operation is simple and it's possible to refactor the method and keep only the stuff we need in light normalization, but I don't want to refactor that until it becomes an issue or we need some performance improvement.

jjbayer commented on tests/integration/test_filters.py:
```diff
@@ -58,6 +58,59 @@ def test_filters_are_applied(
+)
+@pytest.mark.parametrize(
+    "enable_filters, must_filter",
+    ((False, False), (True, True)),
```
nit: One parameter would be enough here.

iker-barriocanal commented on relay-general/src/store/normalize.rs:
```diff
@@ -555,6 +584,7 @@ pub fn light_normalize_event(
         )?; // Timestamps are core in the metrics extraction
         normalize_event_tags(event)?; // Tags are added to every metric
         normalize_exceptions(event)?; // Browser extension filters look at the stacktrace
+        light_normalize_stacktraces(event)?;
```
I thought about including the logic of `light_normalize_stacktraces` inside `normalize_exceptions`. It's a valid option with potentially simpler code, but forces `normalize_exceptions` to only be used as part of the light normalization process, and I don't want to make that assumption when using the method (plus, it makes the method less reusable). Happy to see what others think.

jjbayer replied:
I think keeping them separate is OK. Personally I would move the entire `process_exception` from store normalization to light normalization. We do have to be careful with the order of processing steps though: If `process_exception` used to run before `normalize_exceptions` before we introduced light normalization, it should also run before now.

jjbayer replied:
Looks like `event.process_child_values()` (which calls `process_exception`) used to run before `normalize_exceptions`, so we should the same thing here (https://github.com/getsentry/relay/pull/1366/files#diff-4be114daeb9052778ae699ad5942a07456dd5b519f04d4528404eb2c5a5a2944L507-L568).

iker-barriocanal replied:
The reason not to move the entire `process_exception` to light normalization is that it potentially processes a decent number of things that we don't need at this early stage as light normalization is. The benefit is that it may require fewer changes in the code if we need to do some filtering, but that's a trade-off I'm accepting I guess.

And good catch on the order 👍.

iker-barriocanal committed:
fix normalization order

iker-barriocanal committed:
Merge branch 'master' into iker/fix/filter-browser-ext

iker-barriocanal committed:
Delete duplicated param in test

We don't really need the other param as it always has the same value as
the first one.
